File: hall_of_fame/llddang/[PGS]광물캐기.cpp
#include <iostream>
#include <string>
#include <vector>
#define MAX 2147483647

using namespace std;

int calc(int pick, int mineral){
    if(pick <= mineral) return 1;
    if(pick == 1 || mineral == 1) return 5;
    return 25;
}

int routing(vector<int> picks, vector<int> minerals, int tired, int idx){
    int sum = 0;
    for(auto a : picks) sum += a;
    if(sum == 0 || idx >= minerals.size()) return tired;
    
    int result = MAX, tmp, i, j;
    for(i=0; i<3; i++){
        if(picks[i]){
            picks[i]--;
            tmp = 0;
            
            for(j=idx; j<idx+5 && j < minerals.size(); j++){
                tmp += calc(i, minerals[j]);
            }
            result = min(result, routing(picks, minerals, tired + tmp, j));
            picks[i]++;
        }
    }
    return result;
}

int solution(vector<int> picks, vector<string> minerals) {
    int answer = 0;
    
    vector<int> temp;
    for(auto str : minerals){
        if(str == "diamond") temp.push_back(0);
        if(str == "iron")    temp.push_back(1);
        if(str == "stone")   temp.push_back(2);
    }
    
    answer = routing(picks, temp, 0, 0);
    return answer;
}

File: jaepyo99/[PGS] 광물 캐기.java
import java.util.*;
class Solution {
    static class Mineral {
        private int diamond;
        private int iron;
        private int stone;
        
        public Mineral(int diamond, int iron, int stone) {
            this.diamond = diamond;
            this.iron = iron;
            this.stone = stone;
        }
    }
    static int[][] scoreBoard;
    static List<Mineral> list;
    public int solution(int[] picks, String[] minerals) {
        int answer = 0;
        
        scoreBoard = new int[][]{{1, 1, 1}, {5, 1, 1}, {25, 5, 1}};
        
        int totalPicks = Arrays.stream(picks).sum();
        list = new ArrayList<>();
        for(int i = 0; i < minerals.length; i+=5) {
            if(totalPicks == 0) break;
            
            int dia = 0, iron = 0, stone = 0;
            for(int j = i; j < i + 5; j++) {
                if(j == minerals.length) break;
                
                String mineral = minerals[j];
                int val = mineral.equals("iron") ? 1 : 
                    mineral.equals("stone") ? 2 : 0;
                
                dia += scoreBoard[0][val];
                iron += scoreBoard[1][val];
                stone += scoreBoard[2][val];
            }
            
            list.add(new Mineral(dia, iron, stone));
            totalPicks--;
        }
        
        Collections.sort(list, ((o1, o2) -> (o2.stone - o1.stone)));
        for(Mineral m : list) {
            int dia = m.diamond;
            int iron = m.iron;
            int stone = m.stone;
            
            if(picks[0] > 0) {
                answer += dia;
                picks[0]--;
                continue;
            }
            if(picks[1] > 0) {
                answer += iron;
                picks[1]--;
                continue;
            }
            if(picks[2] > 0) {
                answer += stone;
                picks[2]--;
                continue;
            }
        }
        
        return answer;
    }
}
